src/rust/ide/src/view/node_editor.rs

```diff
@@ -339,20 +355,27 @@ impl GraphEditorIntegratedWithControllerModel {
     }
 
     fn update_node_view
-    (&self, node:graph_editor::NodeId, info:&controller::graph::Node, trees:NodeTrees) {
-        let position = info.metadata.and_then(|md| md.position);
-        if let Some(pos) = position {
-            self.editor.frp.inputs.set_node_position.emit_event(&(node,pos.vector));
+    (&self, id:graph_editor::NodeId, node:&controller::graph::Node, trees:NodeTrees) {
+        let position = node.metadata.and_then(|md| md.position);
+        if let Some(position) = position {
+            self.editor.frp.inputs.set_node_position.emit_event(&(id,position.vector));
         }
-        let expression = info.info.expression().repr();
-        if Some(&expression) != self.expression_views.borrow().get(&node) {
+        let expression = node.info.expression().repr();
+        if Some(&expression) != self.expression_views.borrow().get(&id) {
             let code_and_trees = graph_editor::component::node::port::Expression {
                 code             : expression.clone(),
                 input_span_tree  : trees.inputs,
                 output_span_tree : trees.outputs.unwrap_or_else(default)
             };
-            self.editor.frp.inputs.set_node_expression.emit_event(&(node,code_and_trees));
-            self.expression_views.borrow_mut().insert(node, expression);
+            self.editor.frp.inputs.set_node_expression.emit_event(&(id,code_and_trees));
+            self.expression_views.borrow_mut().insert(id, expression);
+        }
+
+        // Set initially available type information on ports (identifiable expression's sub-parts).
+        for expression_part in node.info.expression().iter_recursive() {
+            if let Some(id) = expression_part.id {
+                self.update_type_on(id)
+            }
```

Comment:
In case of span tree we remember expression and update view only when it changed. But here we notify them each time we got any update from the controllers.

I think we can move this loop inside the if above; we do the full refresh only when expression changes (or node was just put on scene), and do only partial updates when we get notification about new type.